### PR TITLE
proposal: use git submodules for storing imgui dependencies

### PR DESCRIPTION
**Pull request by gucio321**
Hi there!
there is a technology called git submodules that allows to put a git repository inside another one.
I think that instead `Drop source code of imgui to cimgui/imgui` you can just execute `git submodule update --init --recursive`
it has several adventages:
- don't need to search what `imgui` and `implot` exactly are
- clone them with one command
- exact conrol over version (submodule has specified the version that could be set somewhere - i can't remember where :smile: )

### CODE REVIEW

**Comment by AllenDang** [comment]
Thanks! The git action to build also need to change.

**Comment by eliasdaler** [comment]
The disadvantages I see here is that you'll need to generate cimgui outputs manually. 

It's possible to add a GitHub action which will generate and commit outputs on submodule version update. Alternatively, we can add cimgui generation as a build step.

**Comment by gucio321** [comment]
sure, I can write some workflow

**Comment by gucio321** [comment]
[here](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/How-to-update-Git-submodules) is how Am I going to update submodules in workflow ;-)

**Comment by AllenDang** [comment]
@gucio321 Check the ".github/workflows", run them via "Actions" for win/mac/linux, make sure they could be compiled successfully.

**Comment by gucio321** [comment]
@AllenDang I've a question: do I need to regenerate the whole binding after updating dependencies?

**Comment by gucio321** [comment]
ok, it seems to work now
@AllenDang is this workflow ok?
it updates submodules and regenerates binding. if regeneration crashes it'll exit with error
The workflow runs every week on monday

**Comment by eliasdaler** [comment]
@gucio321, can you please manually run `luajit ./generator.lua gcc "comments internal" glfw opengl3 -DIMGUI_DISABLE_OBSOLETE_FUNCTIONS -DIMGUI_USE_WCHAR32` for cimgui generation instead of `generator.sh`?

Will be useful for #27

**Comment by gucio321** [comment]
@eliasdaler like this? 40084d58d6968d9db993724f2642b36085ab42c7

Don't you think that reffering to sh script would be better than hard-codding a command in workflow?

File: .github/workflows/update_submodules.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+          luajit ./generator.lua gcc "comments internal" glfw opengl3 -DIMGUI_DISABLE_OBSOLETE_FUNCTIONS -DIMGUI_USE_WCHAR32
+          just gencode_cimgui
+          cd ../../
+          cd cimplot/generator
+          luajit ./generator.lua gcc "comments internal" glfw opengl3 -DIMGUI_DISABLE_OBSOLETE_FUNCTIONS -DIMGUI_USE_WCHAR32
+          just gencode_cimplot
+          cd ../../
+      - uses: stefanzweifel/git-auto-commit-action@v4
+        with:
+          commit_message: Update dependencies; regenerat code
```
Comment (eliasdaler) [inline .github/workflows/update_submodules.yaml:48]: s/regenerat/regenerate

**Comment by eliasdaler** [comment]
> Don't you think that reffering to sh script would be better than hard-codding a command in workflow?

The problem is that currently generator.sh for cimgui doesn't generate comments because it's called with "internal" not with "comments internal". For cimplot you can still run generator.sh for now.

**Comment by gucio321** [comment]
> The problem is that currently generator.sh for cimgui doesn't generate comments because it's called with "internal" not with "comments internal".

okey, wo why not change this argument in generator.sh?

**Comment by eliasdaler** [comment]
@AllenDang, here's a thing to think about as well... Do we always want cimgui-go to have the latest versions of Dear ImGui and implot? It would be good to have some correspondence between cimgui-go tags and cimgui/cimplot tags, but I'm not sure how this can be implemented.

> okey, wo why not change this argument in generator.sh?
Because it comes from cimgui and it's a bad idea to change files in submodules.

**Comment by eliasdaler** [comment]
Oh, I now see that you only added imgui as a submodule. Why not add cimgui as a submodule instead? I think it makes a bit more sense.

**Comment by gucio321** [comment]
oh, I haven't noticed that cimgui/cimplot are external repos as well ;-)

**Comment by gucio321** [comment]
> @AllenDang, here's a thing to think about as well... Do we always want cimgui-go to have the latest versions of Dear ImGui and implot? It would be good to have some correspondence between cimgui-go tags and cimgui/cimplot tags, but I'm not sure how this can be implemented.

I think you're right, when updating, we should reffer to `latest` tag instead `master` branch

**Comment by eliasdaler** [comment]
With cimgui it's a bit more complicated, because some users might want 1.88 and some might want 1.88dock from cimgui.
Not sure how to handle this. I'd go with 1.88dock personally if it doesn't break much stuff for those who don't use docking.

**Comment by gucio321** [comment]
> Because it comes from cimgui and it's a bad idea to change files in submodules.

is there any reason why cimgui doesn't want it? mayb just create a PR to their repo?

**Comment by gucio321** [comment]
meh, crash...
image
need to figure out how to clone submodules of submodules :simle:

**Comment by eliasdaler** [comment]
> "is there any reason why cimgui doesn't want it? mayb just create a PR to their repo?"
This feature is still not in any tagged release, I guess that's why. We'll see what Allen think about which version is better to use.

@gucio321, you need to do recursive submodule init because cimgui/cimplot have submodules as well.

**Comment by gucio321** [comment]
ok, @eliasdaler now it works, could you please review again

File: .github/workflows/update_dependencies.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          sh generator.sh
+          just gencode_cimgui
+          cd ../../
+          cd cimplot/generator
+          sh generator.sh
+          just gencode_cimplot
+          cd ../../
+      - uses: stefanzweifel/git-auto-commit-action@v4
+        with:
+          commit_message: Update dependencies; tregenerate code
```
Comment (eliasdaler) [inline .github/workflows/update_dependencies.yaml:48]: s/tregenerate/regenerate :smile:

**Review by eliasdaler — COMMENTED** [review]
Looks fine overall, I'll test how well it works with my code a bit later today.

**Comment by eliasdaler** [comment]
@gucio321, I also think that justfile needs some adjustments. It looks like it's currently not possible to do "just gencode_cimgui" locally and build new *.go files
I think that ideally justfile needs to be changed to be able to generate cimgui/cimplot outputs by just running "just gencode_cimgui" and "just gencode_cimplot"

The generation should ideally generate output files to some place *outside* of cimgui/cimplot directories. These directories should also be added to this repo's .gitignore and not commited.

This will also make GitHub workflows easier as you'll be able to just run "just gencode_cimlot" and "just gencode_cimgui".

It would also be great to add ability to detect that user didn't clone submodules before running "just <target>" by checking for some file's existence inside cimgui/cimplot dirs (e.g. README.md)

**Comment by eliasdaler** [comment]
@gucio321 any updates on this? It's almost ready, but needs to work locally :)

**Comment by gucio321** [comment]
oh, nvm, i had corrupted submodules :smile:

> @gucio321 any updates on this? It's almost ready, but needs to work locally :)

so @eliasdaler whot do you think, should be done here? IMO this is complete. it works locally for me

**Comment by eliasdaler** [comment]
Okay, it works for me locally, that's good.
But I have two question:

1. How exactly does that "auto update dependencies" task works? On which branches will it run? It only needs to run on "main" branch, imo
2. It looks like you can drop "sh generator.sh" in workflows because the both cimgui and cimplot already have bindings generated inside of them

**Comment by gucio321** [comment]
> How exactly does that "auto update dependencies" task work? On which branches will it run? It only needs to run on "main" branch, imo. Plus, I think it can run once a month. Dear ImGui do a release every week. And if it does some hotfixes, we can manually do the update.

it, every week, updates cimgui/cimplot deps and regenerates framework, it will run only on master (main)

> It looks like you can drop "sh generator.sh" in workflows because the both cimgui and cimplot already have bindings generated inside of them

it does instruction from readme tbh :smile:
I can remove this step, but, imo, we can regenerate C binding to make sure everything is correct there

**Comment by eliasdaler** [comment]
If you don't want to remove C binding generation, you should make it so that C binding is generated outside of submodule director. Otherwise, you might run into problems if you change files inside submodule directories

**Comment by gucio321** [comment]
> Otherwise, you might run into problems if you change files inside submodule directories

ok, fair enough, removing step responsible for regenerating c binding

**Comment by gucio321** [comment]
@eliasdaler done, now `Update Dependencies` only runs `just`

**Comment by AllenDang** [comment]
@gucio321 @eliasdaler Seems ready to merge?

**Comment by eliasdaler** [comment]
@AllenDang LGTM

**Comment by AllenDang** [comment]
> @AllenDang LGTM

May I ask what LGTM stands for...

**Comment by eliasdaler** [comment]
> > @AllenDang LGTM
> 
> May I ask what LGTM stands for...

"Looks good to me" -> let's get this merged :D

**Pull request merged**